WordPress/Classes/ViewRelated/Views/WPRichTextGallery.swift

```diff
@@ -0,0 +1,335 @@
+                cell.loadingIndicator.stopAnimating()
+            } else {
+                let index = mediaArray.count
+                let indexPath = IndexPath(row: index, section: 1)
+                cell.loadingIndicator.startAnimating()
+                imageSource?.fetchImage(for: url, with: maxDisplaySize, indexPath: indexPath, isPrivate: isPrivate)
+            }
+
+            let media = GalleryMedia(cell: cell, attachment: attachment)
+            mediaArray.append(media)
+        }
+
+        return cell
+    }
+
+}
+
+extension WPRichTextGallery: UICollectionViewDelegateFlowLayout {
+    
+    func setOffsetIfNeeded() {
+        //TODO: Still Needs Fixing
+        let leadingEdge = frame.minX - (superview?.frame.minX ?? 0)
+        
+        if leadingEdge < 16 {
+            let neededOffset = 16 - leadingEdge
+            
+            collectionView.contentInset = UIEdgeInsets(top: 0, left: neededOffset, bottom: 0, right: 0)
+        }
+    }
+
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+
+        let leadingEdge = frame.minX - (superview?.frame.minX ?? 0)
+        let superWidth = superview?.frame.width ?? 0
+        let cellWidth = superWidth - (2 * leadingEdge)
+
+        self.cellWidth = cellWidth
+
+        let height = collectionView.bounds.height > 0 ? collectionView.bounds.height : viewHeight
+
+        return CGSize(width: cellWidth, height: height)
+    }
+
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        handleGalleryTapped?(indexPath.row, imageAttachments)
+    }
+
+    /// Stackoverflow
+    /// https://stackoverflow.com/a/46303794
+    /// In case the user scrolls for a long swipe, the scroll view should animate to the nearest page when the scrollview decelerated.
+    ///
+
+    func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        scrollToPage(scrollView, withVelocity: velocity)
+    }
+
+    func scrollToPage(_ scrollView: UIScrollView, withVelocity velocity: CGPoint) {
+
+        var page = calculateCurrentPage(scrollView: scrollView)
+        if velocity.x > 0 {
+            page += 1
+        }
+        if velocity.x < 0 {
+            page -= 1
+        }
+        page = max(page, 0)
+        let newOffset: CGFloat = CGFloat(page) * (cellWidth + cellPadding)
+
+        scrollView.setContentOffset(CGPoint(x: newOffset, y: 0), animated: true)
+    }
+
+    func calculateCurrentPage(scrollView: UIScrollView) -> Int {
+
+        let page  = Int((scrollView.contentOffset.x - cellWidth / 2) / (cellWidth + cellPadding) + 1)
+        return page
+    }
+
+    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        scrollToPage(scrollView, withVelocity: CGPoint(x: 0, y: 0))
```

Comment:
Is this necessary if `scrollViewWillEndDragging` is implemented and sets an adjusted offset?